Debug-frontend-tests.md

```diff
@@ -0,0 +1,118 @@
+
+Below, we'll discuss some strategies you can use to overcome these difficulties. We suggest you follow a workflow like this:
+
+1. [Find the code for the failing test](#find-tests)
+2. [Run the test in isolation](#isolate-tests)
+3. Debug by [printing out debugging information](#verbose-mode), [investigating the code referenced by the stack trace](#find-stack-elements), and/or [running many times on CI](#run-tests-repeatedly-on-ci).
+
+## Find tests
+
+You can use test names to find test code files, but you can't just search for the whole name. You have to understand how test names are constructed. For example, here's the kind of test structure that produced the test name "Assets Backend API Service on dev mode should handle rejection when fetching a file fail" above:
+
+```js
+describe('Assets Backend API Service', () => {
+  describe('on dev mode', () => }
+    it('should handle rejection when fetching a file fails', () => {
+      ...
+    });
+  });
+});
+```
+
+The strings from the `describe` and `it` functions get concatenated with space (` `) delimiters to create the test name. The name of the `it` block usually starts with `should`, so we recommend searching the codebase for the portion of the test name from `should` to the end (e.g. "should handle rejection when fetching a file fails" in this case).
+
+## Isolate tests
+
+Once you have identified which test is failing, you often want to run that test in isolation next. This will let you get results quickly as you try other debugging strategies. Thankfully, you just have to change the `it()` of your test to `fit()`. You can even do this multiple times to mark multiple tests for running. Then when you run the frontend tests, only those tests with `fit()` will run. Note that `fdescribe()` works the same way.
+
+**Remember to revert your `fit()` and `fdescribe()` changes before committing!**
+
+Once you've specified which tests you want to run, you should make sure you can still reproduce the bug you are investigating. Some bugs are caused by how tests relate to each other, so in these cases, you won't be able to reproduce the bug by running the failing test in isolation. To see how this might happen, consider the following test code:
+
+```js
+describe('oppia', () => {
+  var test = 2;
+
+  it('should do something', () => {
+    test = 3;
+  });
+
+  it('should do something else', () => {
+    expect(test).toBe(2);
+  });
+});
+```
+
+These tests will fail if `should do something` runs before `should do something else`, but not if the tests run in the reverse order. Our frontend tests run in a non-deterministic order, so these kinds of problems generally appear as flakes. In other words, these tests will sometimes pass but fail other times, even if the underlying code is the same.
```

Comment:
We should perhaps talk about trying to avoid such cases and that it is our goal to not have that. If a developer finds such a case should they report it, fix it, etc.?

Reply:
I think we want developers to treat these cases as incorrect tests and fix them. I've added a note to state this